### PR TITLE
Add HDBScan with UMAP

### PR DESCRIPTION
**Pull request by nsthorat**
Uses UMAP to reduce the dimensionality before running HDBScan, greatly improving the performance of clustering.

### CODE REVIEW

File: lilac/signals/cluster_hdbscan.py [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+          all_vectors.append(vector['vector'])
+
+    # Use UMAP to reduce the dimensionality before hdbscan to speed up clustering.
+    # For details on hyperparameters, see:
+    # https://umap-learn.readthedocs.io/en/latest/clustering.html
+    reducer = umap.UMAP(n_components=10, n_neighbors=30, min_dist=0.0)
+    all_vectors = reducer.fit_transform(all_vectors)
+
+    hdbscan = HDBSCAN(min_cluster_size=self.min_cluster_size, n_jobs=-1)
+    hdbscan.fit(np.array(all_vectors))
```
Comment (dsmilkov) [inline lilac/signals/cluster_hdbscan.py:68]: nit: no need to wrap in np.array() since all_vectors comes from umap

File: lilac/signals/cluster_hdbscan.py [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+    with DebugTimer('DBSCAN: Reading from vector store'):
+      for vectors in span_vectors:
+        all_spans.append([vector['span'] for vector in vectors])
+        for vector in vectors:
+          all_vectors.append(vector['vector'])
+
+    # Use UMAP to reduce the dimensionality before hdbscan to speed up clustering.
+    # For details on hyperparameters, see:
+    # https://umap-learn.readthedocs.io/en/latest/clustering.html
+    reducer = umap.UMAP(n_components=10, n_neighbors=30, min_dist=0.0)
```
Comment (dsmilkov) [inline lilac/signals/cluster_hdbscan.py:64]: extract n_components into a constant since it's the param that balances speed/accuracy

File: lilac/signals/cluster_hdbscan.py [hunk 3]
```diff
@@ -0,0 +1,81 @@
+"""Compute clusters for a dataset."""
+from typing import ClassVar, Iterable, Optional
+
+import numpy as np
+import umap
```
Comment (dsmilkov) [inline lilac/signals/cluster_hdbscan.py:5]: since we made umap-learn optional , import it in a try/catch inside compute().

Alternatively, and maybe better is to make umap-learn required since we will probably use it in "Cluster View" when we add 2d plots

**Review by dsmilkov — APPROVED** [review]
